### PR TITLE
Module deprecation: docs, scheme and tests

### PR DESCRIPTION
**Pull request by gundalow**
##### SUMMARY
Enforce module deprecation.
After module has reached the end of it's deprecation cycle we will replace it with a docs stub.

* Use of deprecated past deprecation cycle gives meaningful message (see examples below)
* Enforce documentation.deprecation dict via `schema.py`
* Update `ansible-doc` and web docs to display documentation.deprecation
* Document that structure in `dev_guide`
* Ensure that all modules starting with `_` have a `deprecation:` block
* Ensure `deprecation:` block is only used on modules that start with `_`
* `removed_in` A string which represents when this module needs **deleting**
* CHANGELOG.md and porting_guide_2.5.rst list removed modules as well as alternatives
* CHANGELOG.md links to porting guide index

Ensure that if the module is now at the end of it's deprecation it contains:
```python
if __name__ == '__main__':
    removed_module()
```

Remove code from the following modules (some of which should have been in in 2.4 and earlier:
* network/nxos/_nxos_mtu.py
* network/cumulus/_cl_interface_policy.py
* network/cumulus/_cl_bridge.py
* network/cumulus/_cl_img_install.py
* network/cumulus/_cl_ports.py
* network/cumulus/_cl_license.py
* network/cumulus/_cl_interface.py
* network/cumulus/_cl_bond.py
* cloud/amazon/_ec2_vpc.py
* cloud/amazon/_ec2_ami_search.py
* cloud/docker/_docker.py
* utilities/helper/_accelerate.py
* plus unit and integration tests for the above

**Example of `_foo.py` without `deprecation:`**
```
ERROR: lib/ansible/modules/network/nxos/_foo.py:0:0: E305 DOCUMENTATION.deprecated: required key not provided @ data['deprecated']. Got None
ERROR: lib/ansible/modules/network/nxos/_foo.py:0:0: E305 DOCUMENTATION.module: not a valid value for dictionary value @ data['module']. Got 'nxos_ip_interface'
ERROR: lib/ansible/modules/network/nxos/_foo.py:0:0: E318 Module deprecated, but DOCUMENTATION.deprecated is missing
```
**Example of module past deprecation cycle, though with code in**
```
lib/ansible/modules/network/nxos/_nxos_mtu.py:0:0: E103 Did not find a call to removed_module()
```

**ansible-doc**
`ansible-doc kubernetes`

```
DEPRECATED: 

        Reason: This module used the oc command line tool, where as M(k8s_raw) goes over the REST API.
        Will be removed in: Ansible 2.9
        Alternatives: Use M(k8s_raw) instead.

```

**Before removed_module()**
```
ansible localhost -m accelerate
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

[DEPRECATION WARNING]: accelerate is kept for backwards compatibility but usage is discouraged. The module documentation details page may explain more about this rationale.. This feature will be removed in a future release. Deprecation 
warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
localhost | FAILED! => {
    "changed": false, 
    "failed": true, 
    "module_stderr": "", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE", 
    "rc": 0
}
```

**With removed_module()**
Note sensible `msg`
```
ansible localhost -m accelerate
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

[DEPRECATION WARNING]: accelerate is kept for backwards compatibility but usage is discouraged. The module documentation details page may explain more about this rationale.. This feature will be removed in a future release. Deprecation 
warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
localhost | FAILED! => {
    "changed": false, 
    "failed": true, 
    "msg": "This module has been removed.  The module documentation may contain hints for porting"
}
```
**NOTE** Future PR will be raised to ensure we don't get a deprecation warning followed by "this has been removed". Not sure what the correct solution to that would be, may come after 2.5.

##### ISSUE TYPE
 - Bugfix Pull Request
 - Docs Pull Request

### CODE REVIEW

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst [hunk 1]
```diff
@@ -202,9 +202,11 @@ The following fields can be used and are all required unless specified otherwise
 :author:
   Name of the module author in the form ``First Last (@GitHubID)``. Use a multi-line list if there is more than one author.
 :deprecated:
-  If this module is deprecated, detail when that happened, and what to use instead, e.g.
-  `Deprecated in 2.3. Use M(whatmoduletouseinstead) instead.`
-  Ensure `CHANGELOG.md` is updated to reflect this.
+  In addition to updating the ``CHANGELOG`` and ``porting_guide.X.y.rst`` the module should be renamed to start with an ``_`` and the following values should be set:
+
+  :version: A `string` which represents the current version of Ansible, i.e. ``version: "2.5"``. Not the version this will be removed in.
```
Comment (abadger) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: "represents the version of Ansible the deprecation starts"

File: lib/ansible/modules/cloud/amazon/_ec2_ami_search.py [hunk 2]
```diff
@@ -16,7 +16,9 @@
 ---
 module: ec2_ami_search
 short_description: Retrieve AWS AMI information for a given operating system.
-deprecated: "Use M(ec2_ami_find) instead."
+deprecated:
+  version: "2.4"
```
Comment (abadger) [inline lib/ansible/modules/cloud/amazon/_ec2_ami_search.py]: This goes back a lot further.  I believe it's 2.0: https://github.com/ansible/ansible-modules-core/tree/stable-2.0/cloud/amazon

**Comment by abadger** [comment]
A few comments but the rest looks good.

File: lib/ansible/modules/cloud/amazon/_ec2_ami_search.py (hunk 2 above)
Comment (gundalow) [reply lib/ansible/modules/cloud/amazon/_ec2_ami_search.py]: Yup, so I think this should be deleted

**Comment by gundalow** [comment]
@abadger Thanks. I've added `2.0` into the allow list. Maybe a candidate for deleting at the beginning of 2.6?

**Comment by mattclay** [comment]
CI failed sanity test with:

```
ERROR! module accelerate missing documentation (or could not parse documentation): 'why'
```

The bot should have reported this (and may do so after my comment).

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst [hunk 3]
```diff
@@ -202,9 +202,11 @@ The following fields can be used and are all required unless specified otherwise
 :author:
   Name of the module author in the form ``First Last (@GitHubID)``. Use a multi-line list if there is more than one author.
 :deprecated:
-  If this module is deprecated, detail when that happened, and what to use instead, e.g.
-  `Deprecated in 2.3. Use M(whatmoduletouseinstead) instead.`
-  Ensure `CHANGELOG.md` is updated to reflect this.
+  In addition to updating the ``CHANGELOG`` and ``porting_guide.X.y.rst`` the module should be renamed to start with an ``_`` and the following values should be set:
```
Comment (mattclay) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: We should also mention changing the status metadata.

**Review by mattclay — CHANGES_REQUESTED** [review]

**Comment by gundalow** [comment]
Ah, so `ansible-doc` requires `version`, `why` and `alternative` to be set. So I'll work through and flesh this out.

It's possible `ansible-doc` has the wrong message in:
`Scheduled removal: Ansible %(version)`, I think that should be: deprecated in.

**Review by mattclay — APPROVED** [review]

**Comment by gundalow** [comment]
My current implementation treats `version` as "notice of deprecation given as of version"

File: test/sanity/validate-modules/schema.py [hunk 4]
```diff
@@ -83,23 +83,51 @@ def return_schema(data):
...
 
 
+def deprecation_schema():
+
+    deprecation_schema_dict = {
+        # Every time we branch:
+        #   * old versions should be removed from the beginning of this list
+        #   * git rm old modules
+        #   * Update CHANGELOG & porting_guide
+        Required('version'): Any("2.0", "2.2", "2.3", "2.4", "2.5"),
```
Comment (sivel) [inline test/sanity/validate-modules/schema.py]: Should we really enumerate all of these versions here?  Maybe it should be current-5 through current?

Maybe used/calculated like:

```python
from ansible import __version__ as ansible_version
ansible_release_version = float('.'.join(ansible_version.split('.')[:2]))
deprecation_range = ['%0.1f' % ((ansible_release_version-0.4) + float(x/10.0)) for x in range(0, 6)]
Required('version'): Any(*deprecation_range)
```

That would produce something like:

```python
['2.1', '2.2', '2.3', '2.4', '2.5', '2.6']
```

That is basically current-4 through current+1, and could of course be adjusted.

File: test/sanity/validate-modules/schema.py [hunk 5]
```diff
@@ -83,23 +83,51 @@ def return_schema(data):
...
+        Required('why'): Any(*string_types),
+        Required('alternative'): Any(*string_types),
+        'removed': Any(True),
```
Comment (sivel) [inline test/sanity/validate-modules/schema.py]: I don't think we need a `removed=True`.  With docs being versioned, there should never be a need to keep that info.

File: lib/ansible/modules/utilities/helper/_accelerate.py [hunk 6]
```diff
@@ -16,11 +16,14 @@
 DOCUMENTATION = '''
 ---
 module: accelerate
-removed: True
```
Comment (sivel) [inline lib/ansible/modules/utilities/helper/_accelerate.py:19]: I'm not sure there is good reason to keep this any longer.  We have versioned docs now, so we should just `git rm` this.

**Review by sivel — DISMISSED** [review]

File: lib/ansible/modules/utilities/helper/_accelerate.py (hunk 6 above)
Comment (gundalow) [reply lib/ansible/modules/utilities/helper/_accelerate.py:19]: bcoca said this file is complex and we don't want to remove it yet.

File: test/sanity/validate-modules/schema.py (hunk 5 above)
Comment (gundalow) [reply test/sanity/validate-modules/schema.py]: Agreed, removed from here and the only module it's been used in

File: test/sanity/validate-modules/schema.py (hunk 4 above)
Comment (gundalow) [reply test/sanity/validate-modules/schema.py]: Nice, though not sure at the moment

**Comment by gundalow** [comment]
@sivel made the good point that the have docs somewhere that indicate the deprecation cycle.


How about in the webdocs, if a module is deprecated we link to the porting guide index, which should also be updated to link to the deprecation & release cycle pages

**Comment by gundalow** [comment]
Added to Core agenda https://github.com/ansible/community/issues/291#issuecomment-355958604

File: lib/ansible/modules/utilities/helper/_accelerate.py [hunk 7]
```diff
@@ -16,11 +16,14 @@
...
 short_description: Enable accelerated mode on remote node
-deprecated: "Use SSH with ControlPersist instead."
+deprecated:
+  removed_in: "2.4"
+  why: Replaces by ControlPersist
```
Comment (abadger) [inline lib/ansible/modules/utilities/helper/_accelerate.py]: s/replaces/Replaced/

File: lib/ansible/modules/utilities/logic/_include.py [hunk 8]
```diff
@@ -20,8 +20,9 @@
 module: include
 short_description: Include a play or task list
 deprecated:
-  The include action was too confusing, dealing with both plays and tasks, being both dynamic and static. This module
-  will be removed in version 2.8. As alternatives use M(include_tasks), M(import_playbook), M(import_tasks).
+  removed_in: "2.8"
+  why: The include action was too confusing, dealing with both plays and tasks, being both dynamic and static. This module will be removed in version 2.8.
+  alternative: Use M(include_tasks), M(import_playbook), M(import_tasks).
```
Comment (abadger) [inline lib/ansible/modules/utilities/logic/_include.py:25]: Should this list of alternatives be expanded?

File: lib/ansible/modules/network/nxos/_nxos_mtu.py [hunk 9]
```diff
@@ -25,7 +25,10 @@
 module: nxos_mtu
 extends_documentation_fragment: nxos
 version_added: "2.2"
-deprecated: Deprecated in 2.3 use M(nxos_system)'s C(mtu) option.
+deprecated:
+  removed_in: "2.3"
```
Comment (abadger) [inline lib/ansible/modules/network/nxos/_nxos_mtu.py]: I still see code in this file.  SHould it be removed_in: "2.7" ?

File: lib/ansible/modules/network/cumulus/_cl_ports.py [hunk 10]
```diff
@@ -19,7 +19,10 @@
 version_added: "2.1"
 author: "Cumulus Networks (@CumulusNetworks)"
 short_description: Configure Cumulus Switch port attributes (ports.conf)
-deprecated: Deprecated in 2.3. Use M(nclu) instead.
+deprecated:
+  removed_in: "2.3"
```
Comment (abadger) [inline lib/ansible/modules/network/cumulus/_cl_ports.py]: I still see code in this file. Should it be removed_in: "2.7" ?

File: lib/ansible/modules/utilities/helper/_accelerate.py (hunk 7 above)
Comment (gundalow) [reply lib/ansible/modules/utilities/helper/_accelerate.py]: Good spot, fixed

File: lib/ansible/modules/utilities/logic/_include.py (hunk 8 above)
Comment (gundalow) [reply lib/ansible/modules/utilities/logic/_include.py:25]: Jimi said no

File: lib/ansible/modules/network/nxos/_nxos_mtu.py (hunk 9 above)
Comment (gundalow) [reply lib/ansible/modules/network/nxos/_nxos_mtu.py]: Reviewing these by hand again

File: lib/ansible/modules/network/cumulus/_cl_ports.py (hunk 10 above)
Comment (gundalow) [reply lib/ansible/modules/network/cumulus/_cl_ports.py]: Reviewing these by hand again

**Comment by gundalow** [comment]
So I think this is good for final review now
* Review and merge this
* Separate PRs to `git rm` old modules
* Separate PR to create deprecation docs, which will be linked from porting guide, and from deprecated modules (may not be done for 2.5)

**Comment by mattclay** [comment]
CI failure in unit tests:

```
test/units/modules/cloud/docker/test_docker.py:5: in <module>
    from ansible.modules.cloud.docker._docker import get_split_image_tag
E   ImportError: No module named _docker
```

**Comment by gundalow** [comment]
Should be good
ready_for_review

**Review by sivel — APPROVED** [review]

File: lib/ansible/modules/network/nxos/_nxos_mtu.py [hunk 11]
```diff
@@ -25,7 +25,10 @@
 module: nxos_mtu
 extends_documentation_fragment: nxos
 version_added: "2.2"
-deprecated: Deprecated in 2.3 use M(nxos_system)'s C(mtu) option.
+deprecated:
+  removed_in: "2.5"
+  why: Replaced with common C(*_system) network modules.
+  alternative: Use M(nxos_system)'s C(mtu) option.
```
Comment (trishnaguha) [inline lib/ansible/modules/network/nxos/_nxos_mtu.py]: `alternative` should be `Use M(nxos_system)'s C(system_mtu) option.`

**Review by trishnaguha — DISMISSED** [review]

**Comment by gundalow** [comment]
bot_status

**Comment by gundalow** [comment]
bot_status

**Review by trishnaguha — APPROVED** [review]

**Comment by gundalow** [comment]
Any open issues or PRs for the above modules that are "docs only stubs" have been closed

**Pull request merged**